### PR TITLE
 [SYCL][CUDA] Expose context extended deleters on PI API

### PR DESCRIPTION
**Pull request by StuartDAdams**
This PR replaces the PI CUDA backend's context callback mechanism with a PI API extension. Users of PI API can now register function pointers to be called before the context is destroyed. Prior to this PR, registering a function required including pi_cuda.hpp. Now that pi_cuda.hpp is private and not accessible to external projects, exposing this functionality on public PI API has become necessary.

### CODE REVIEW

File: sycl/source/detail/pi.cpp [hunk 1]
```diff
@@ -53,6 +55,16 @@ namespace pi {
 
 bool XPTIInitDone = false;
 
+void contextSetExtendedDeleter(cl::sycl::context context,
+                               pi_context_extended_deleter func,
+                               void *user_data) {
+  auto impl = getSyclObjImpl(context);
+  auto contextHandle = reinterpret_cast<pi_context>(context.get());
```
Comment (romanovvlad) [inline sycl/source/detail/pi.cpp:62]: Please, use `getHandleRef`. get increments reference count(piRetainContext).

File: sycl/include/CL/sycl/detail/pi.hpp [hunk 2]
```diff
@@ -95,6 +98,10 @@ using PiMemObjectType = ::pi_mem_type;
 using PiMemImageChannelOrder = ::pi_image_channel_order;
 using PiMemImageChannelType = ::pi_image_channel_type;
 
+void contextSetExtendedDeleter(cl::sycl::context context,
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/pi.hpp:101]: Can we take context by const ref?

**Comment by bader** [comment]
@StuartDAdams, please, fix failed checks.

File: sycl/source/detail/pi.cpp (hunk 1 above)
Comment (romanovvlad) [reply sycl/source/detail/pi.cpp:62]: Could you please resolve this one?
```suggestion
  auto contextHandle = reinterpret_cast<pi_context>(impl->getHandleRef());
```

File: sycl/source/detail/pi.cpp [hunk 3]
```diff
@@ -53,6 +55,18 @@ namespace pi {
...
+void contextSetExtendedDeleter(cl::sycl::context context,
+                               pi_context_extended_deleter func,
+                               void *user_data) {
+  auto impl = getSyclObjImpl(context);
+  auto contextHandle = reinterpret_cast<pi_context>(impl->getHandleRef());
+  auto plugin = impl->getPlugin();
+  plugin.call_nocheck<PiApiKind::piextContextSetExtendedDeleter>(
+      contextHandle, func, user_data);
+  plugin.call_nocheck<PiApiKind::piContextRelease>(
+      contextHandle);
```
Comment (romanovvlad) [inline sycl/source/detail/pi.cpp:67]: Please, remove this as you do not call call "get" anymore.

File: sycl/include/CL/sycl/detail/pi.hpp (hunk 2 above)
Comment (StuartDAdams) [reply sycl/include/CL/sycl/detail/pi.hpp]: Done

File: sycl/source/detail/pi.cpp (hunk 3 above)
Comment (StuartDAdams) [reply sycl/source/detail/pi.cpp]: DOne

**Pull request merged**